patches/0012-add-support-for-logging-used-Windows-APIs.patch

```diff
@@ -0,0 +1,372 @@
++		if err := out.Close(); err != nil {
++			log.Fatal(err)
++		}
++	}()
++	out.WriteString("// Code generated by syscalltrace_generate_windows.go. DO NOT EDIT.\n\n")
++	out.WriteString("package runtime\n\n")
++	out.WriteString("var staticSyscalls = [...]string{\n")
++	for _, file := range flag.Args() {
++		content, err := os.ReadFile(file + ".go")
++		if err != nil {
++			log.Fatal(err)
++		}
++		scanner := bufio.NewScanner(bytes.NewReader(content))
++		for scanner.Scan() {
++			if after, found := strings.CutPrefix(scanner.Text(), "//go:cgo_import_dynamic runtime._"); found {
++				name, _, _ := strings.Cut(after, " ")
++				out.WriteString("\t\"" + name + `\x00"` + ",\n")
++			}
++		}
++	}
++	out.WriteString("}\n")
++}
+diff --git a/src/runtime/syscalltrace_windows.go b/src/runtime/syscalltrace_windows.go
+new file mode 100644
+index 00000000000000..d27e1ffeb9bffe
+--- /dev/null
++++ b/src/runtime/syscalltrace_windows.go
+@@ -0,0 +1,124 @@
++//go:generate go run syscalltrace_generate_windows.go -output zsyscalltrace_windows.go os_windows syscalltrace_windows
++
++package runtime
++
++import "unsafe"
++
++//go:cgo_import_dynamic runtime._CreateFileW CreateFileW%7 "kernel32.dll"
++//go:cgo_import_dynamic runtime._GetEnvironmentVariableW GetEnvironmentVariableW%3 "kernel32.dll"
++//go:cgo_import_dynamic runtime._GetModuleFileNameA GetModuleFileNameA%3 "kernel32.dll"
++
++var (
++	_CreateFileW,
++	_GetEnvironmentVariableW,
++	_GetModuleFileNameA,
++	_ stdFunction
++)
++
++var (
++	syscallTraceHandler uintptr
++	currentProcessName  [_MAX_PATH]byte
++)
++
++// syscallTrace logs a syscall trace.
++// The trace is written as a comma-separated list of the process name,
++// module name, procedure name, and a newline character.
++// If the underlying file supports atomic writes, the trace is guaranteed
++// to be written atomically.
++func syscallTrace(modHandle uintptr, procName *byte) {
++	if syscallTraceHandler == _INVALID_HANDLE_VALUE {
++		return
++	}
++	if syscallTraceHandler == 0 {
++		// It is safe to call initSyscallTrace without locking
++		// because the first time syscallTrace is called is during
++		// runtime initialization, when there is only one goroutine.
++		if !initSyscallTrace() {
++			return
++		}
++	}
++	// We can't allocate memory here, as syscallTrace might be called
++	// before the memory allocator is initialized. Instead, we use a
++	// fixed-size buffer on the stack.
++	buf := make([]byte, _MAX_PATH*3+3)
++	n := findnull(&currentProcessName[0])
++	copy(buf, currentProcessName[:n])
++	buf[n] = ','
++	n += 1
++	// Use GetModuleFileNameA instead of GetModuleFileNameW to avoid
++	// having to allocate memory for the UTF-16 string.
++	n1 := stdcall3(_GetModuleFileNameA, modHandle, uintptr(unsafe.Pointer(&buf[n])), _MAX_PATH)
++	if n1 == 0 || n1 == _MAX_PATH {
++		// Failed or short buffer.
++		buf[n] = '?'
++		n += 1
++	} else {
++		n += int(n1)
++	}
++	buf[n] = ','
++	n += 1
++	procNameLength := findnull(procName)
```

Comment:
Is `procName` (always?) null terminated? Based on `syscall_getprocaddress` accepting `*byte` it certainly seems like it, but figured it might be worth asking.

Reply:
Yep, always, else we would have to pass the string length around. Anyway, I'm adding a comment on the function docs.